
The current plan is ...

Four lectures each:
BG: The basic skill set for practical bioinformatics research (with NGS data)
    UNIX cmd line
         redirection
         pipelines, etc
    Editor - nano
    Programming - R *(maybe to ID?)*
    File formats - fastq, gff
    Quality exploration - fastqc
    Trimmomatic
    Hash tables, SA arrays, BWT, FM datastructures
    Aligners - sam file format

BE: Identifying and utilizing polymorphism data from Illumina resequencing (and reduced representation sequencing)
    Macaque data
    GATK
    RADseq, GBS, etc.
    SNP calling, genotyping
    vcftools, bcf file formats
    base recalibration
    haplotype callers
    bed file format

ID: Introduction to using RNAseq for the analysis of differentially expressed genes 
    0. Markdown (this probably belongs with R) 
    1. Transcriptome Assembly using Trinity (ID can do the practical part: Maybe BG for theory?), and how to interpret and use the **de novo* transcriptome assembly. 
    2. Considerations for types of features to use (genes, transcripts, kmers)
    How to count for RNAseq (what features to use, different approaches to deal with multi-mapped reads)
    3. Basic Statistics for the analysis of count data for differential expression
    4. Differential expression analysis. Examining data for technical issues. Dealing with multiple comparisons (FDR). 
    
