# Bio720

*Please note: this page is currently a stub with minimal information!!*

## Course Summary
This is the page for Bio720, a practical introduction to bioinformatics and genomics research. 

## Background assumed for students
For this class we are not assuming students have background in programming/scripting, nor in bioinformatics. We do assume that students have a working knowledge of basic molecular biology and genetics and have basic familiarity with using computers.

# What students will need

## Course goals
The primary goal of this course it to provide graduate students new to bioinformatics and genomics a practical, basic introduction to the computational skills that will enable them to begin the analysis of genomics data. 
## Learning Objectives

# Topics (mostly TBD at this point)
1. Introduction to UNIX and the command line
2. Transcriptome Assembly
3. Syntenic mapping (Mapping reads to genomes)
4. Identifying and using polymorphism data from NGS data
5. Introduction to Programming in `R` (with applications for using RNASeq data for differential expression analysis and gene expression profiling)
6. RNASeq analysis for differential expression analysis

## Instructors
[Dr. Brian Golding](http://helix.biology.mcmaster.ca/)

[Dr. Ben Evans](https://benevanslab.wordpress.com/)

[Dr. Ian Dworkin](https://scholar.google.com/citations?user=Iium3AEAAAAJ&hl=)

## Recommended books. 
You are responsible for ordering your own copies of these books. Both are excellent, but we are only highly recommending the first book (BDS) for this class.

[Bioinformatics Data Skills](http://www.amazon.ca/Bioinformatics-Data-Skills-Reproducible-Research/dp/1449367372/ref=sr_1_1?s=books&ie=UTF8&qid=1440614667&sr=1-1&keywords=bioinformatics+data+skills). **HIGHLY RECOMMENDED** This is a new book, but which fills an important gap in that is oriented towards the day to day skills for anyone working in the fields of genomics and bioinformatics. In addition to covering the basic UNIX skills (and why we use UNIX in bioinformatics and genomics), it also covers subjects like overviews of the essential file types (`.fasta`, `.fastq`, `.gff`, etc) that are ubiquitous in the field. There is also a nice, but brief introduction to the essentials of `R`, using bioconductor and in particular range data, and two important chapters on how to organizing (and maximize reproducibility) of computational projects. Currently this is ~52$ on Amazon. We have also contact with the publisher to see if we can discount this at all. It is available as an [e-book as well from the publisher](http://shop.oreilly.com/product/0636920030157.do).

[Practical Computing for Biologists](http://www.amazon.com/s/ref=nb_sb_ss_c_0_24?url=search-alias%3Dstripbooks&field-keywords=practical+computing+for+biologists&sprefix=practical+computing+for+biologists%2Caps%2C144). This book provides a nice, gentle introduction to the basic computational skills all biologists should have. In particular, with introduction to using the UNIX command line, shell scripting, basic Python programming, regular expressions, working on remote machines and a few other topics. The book is written to be agnostic with respect to discipline (i.e. it is not a bioinformatics book per se), but does a great job of being both very accessible and immediately useful. It seems a bit pricy on Amazon.ca, but look around for used copies (it is 4 years old). This is an excellent book, but given the current pricing for it (it used to be about 50$) we are suggesting BDS for the course. However, if you plan to continue in computational research, this is a fantastic resource.





