---
title: "Bio720_MuckingWithDrosophilaPromoters"
author: "Ian Dworkin"
date: "12/06/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

# A little introduction to Bioconductor using *Drosophila*


## A visit to Bioconductor
You have to give the developers of Bioconductor a little love. They have really gone out of their way to provide very powerful *R* libraries, with lots of tutorials. Let's take a look shall we. The bioconductor home page is [here](http://bioconductor.org/).

## A little introduction with Drosophila

With 2083 (as of December 6th, 2021) [packages](https://www.bioconductor.org/packages/release/BiocViews.html#___Software) it is not like I can show you everything. However the Bioinformatic Data Skills book, and the datacamp tutorial will give you a tiny bit of an intro into things.

We will just use a few packages today

```{r}
#library(BiocInstaller) # old deprecated
library(BiocManager) # new way
library(Biostrings)
library(GenomicRanges)
library(GenomicFeatures)
library(BSgenome)

library(seqLogo)
```


## old vs new

In terms of installing BioConductor genomes and libraries they have deprecated some functions about 3 years ago. I only mention in because you may see some old code using the `biocLite` function which is now **deprecated**.

** deprecated, so don't use.. BUT YOU WILL SEE THIS IN THE DATACAMP MODULE**
```{r}
#source("https://bioconductor.org/biocLite.R") # deprecated
#biocLite('BSgenome') # deprecated
```

**New!**

```{r}
BiocManager::install("BSgenome") 
```

You only need one, and depending on what version you have, it should be easy to figure out, however, you will want to migrate to the newer version soon.

### Downloading the Drosophila genome

First let us take a look at available genomes

```{r}
available.genomes()
```

This is of course only a handful of genomes to play with, and getting your own genome of interest installed is something for you to do. We with work with a recent version of the *Drosophila melanogaster* genome assembly

```{r}
#biocLite("BSgenome.Dmelanogaster.UCSC.dm6") # Should not need this now!
BiocManager::install("BSgenome.Dmelanogaster.UCSC.dm6")



library(BSgenome.Dmelanogaster.UCSC.dm6)
Dm6 <- BSgenome.Dmelanogaster.UCSC.dm6


BSgenome.Dmelanogaster.UCSC.dm6
Dm6


class(BSgenome.Dmelanogaster.UCSC.dm6)
typeof(BSgenome.Dmelanogaster.UCSC.dm6) # an S4 object

str(BSgenome.Dmelanogaster.UCSC.dm6)  # not so useful from a bioinformatics analysis perspective
 
```



**Question:** Pick another genome and download it and give it initials (like Hs for Homo sapiens)


### Note about Non-model species genomes

There is a pretty reasonable chance that the exact genome you want to work with is not available. There are PLENTY of resources in bioconductor to help you out. In particular check out the [how to guide for the annotation hub](https://bioconductor.org/packages/release/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub-HOWTO.html) and also see here for a [useful vignette](https://www.bioconductor.org/packages/release/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html).

## Some basic statistics about the Drosophila genome

Let's start by just getting some basic information about the genome we are looking at.

```{r}
length(Dm6)  # this is how many "chromosomes" (usually contigs) in this genome

names(Dm6)

length(names(Dm6)) # same as length of the genome BSgenome object

```

**Question** How is it that the length function works on such a "strange" object like Dm6?




For Bioconductor you can get a useful list of the S4 generic functions by looking at 

```{r}
?BiocGenerics
```


Obviously a lot more information available for our genome

```{r}
show(Dm6)
organism(Dm6)

isS4(Dm6)
```


```{r}
seqlengths(Dm6) # length of each contig.
```

**Question:** How would you sort this so that the the largest contigs (representing near complete chromosomal assemblies) are first?




Thankfully there is an easier way of looking at this!

```{r}
seqinfo(Dm6)
```


**Question:** How long is the assembled genome in base pairs (bp)? How about in mega base pairs (Mb)?




## frequency of bases in the genome.

Often we may need to start with something simple like looking at the nucleotide composition of the genome. 


Let's take a look at just one arm of one chromosome.


```{r}
alphabetFrequency(Dm6$chr2R)
```


**Question:** How would you represent this as probabilities instead. Use the help file for alphabetFrequency.

**Question:** Why is there more than just ACGT?



## Single chromosome arm

Let's take a look at just one arm of one chromosome.





We can also use the `getSeq` function to extract a set of sequences, like a chromosome

```{r}

dm_chrX <- getSeq(Dm6, names = "chrX")
dm_chrX

show(dm_chrX)

print(dm_chrX)


class(dm_chrX) # This is now a DNAString type object, not a BSgenome object
typeof(dm_chrX)


length(dm_chrX)/1e6  # hey, why does this work!?   ;)

nchar(dm_chrX)/1e6
```

## Frequencies of nucleotides on the X chromosome
```{r}
alphabetFrequency(dm_chrX, baseOnly = T, as.prob = T)
```



**Question:** How would you extract positions 1000000 to 2000000 examining the negative strand of chromosome 3L? Use the help function for ?getSeq to help you. Then take a look at the nucleotide frequencies.




## Finding some simple patterns in our chromosome

Often we are looking through genomes for certain patterns (usually best set up as a kinf of regular expression, but with the ambiguous nucleotide characters)

```{r}
seqs <- c("CCCGGG", "CCCCTTTT", "AAATTT", "TAGCT")

n = sapply(seqs, function(x) countPattern(x, dm_chrX ))


n

which.max(n)
```


**Question:** Find the same patterns as above, but where the 4th nucleotide in the string can be any nucleotide. Note you will want to check the help file for one of the arguments.




## patterns

In BioStrings, there are a few functions that can really help! Also see the help files.

  - `matchPattern(pattern, subject, max.mismatch = 1)` one string to one string
  - `vmatchPattern(pattern, subject)`
    - one set of strings to one string
    - one string to one set of strings.
  - `findPalindromes` - finds palindromic regions in a single string.
 
 Other tools for PWMs, dictionaries etc..(?`vcountPDict,BSgenome-method`)
 - see `?PWM` and also `matchLRPatterns`


## a position weight matrix for a Drosophila transcription factor HNF4alpha

THis is a collection of known transcription factor (TF) binding sites in the Drosophila genome for the HNF transcriptiona factor that is already in bioconductor (although you may have to install)


```{r}
data(HNF4alpha)
HNF4alpha
class(HNF4alpha)
```


We can generate a position frequency matrix and position weight matrix of these TF binding sites.

```{r}
pfm <- consensusMatrix(HNF4alpha)

pfm
```


Not the most useful way of looking at it. So let's clean this up a bit. 

```{r}

round(pfm[1:4,]/colSums(pfm[1:4,]), 2)
class(pfm)

```


We can now use this and create a position weight matrix (PWM) object

```{r}


pwm <- PWM(pfm)

# or directly

```



While this is useful to see, we can do this more directly (protect those kittens)


```{r}

pwm <- PWM(HNF4alpha, type = "prob")
round(pwm[1:4,]/colSums(pwm[1:4,]), 2)
maxWeights(pwm)

conMat <- consensusMatrix(HNF4alpha, as.prob = T)[1:4,]


seqLogo(conMat)

seqLogo(pwm[1:4,]/colSums(pwm[1:4,]))
```

## matching these to the X chromosome

```{r}
TF_hits <- matchPWM(pwm, dm_chrX, with.score = TRUE)
length(TF_hits)
head(mcols(TF_hits)$score)

hist(mcols(TF_hits)$score)
```

## How about the reverse complement of the X?

```{r}
TF_hits_rc <- matchPWM(reverseComplement(pwm), dm_chrX, with.score = TRUE)
length(TF_hits_rc)
hist(mcols(TF_hits_rc)$score)
```

## Extracting upstream sequences from Drosphila melanogaster

Makes a TxDb object (transcript annotations. Again, lots of options depending on how you want to refer to genes. Then extract the upstream sequences from the genes.
```{r}
dm_txdb <- makeTxDbFromUCSC("dm6", tablename="refGene") 

up10000seqs <- extractUpstreamSeqs(dm, dm_txdb, width=10000)
```

How about this pattern?

```{r}
upstream_hits <- vcountPattern("NGNNCAAAGNNCA", up10000seqs)
length(upstream_hits)
```


