---
title: "Differential Expression analysis using RNA-Seq data with DESeq2 (HTSeq counts)"
author: "Ian Dworkin"
date: "February 8, 2017"
output:
  html_document:
    keep_md: yes
---
# Modified from my NGS2015 Tutorial on Differential Expression analysis using RNA-Seq data and DESeq2

Please examine the tutorial for the count data for salmon on instructions on installation of libraries etc..

If you have already run the other tutorial in the currently running session of `R`, then you will not need to re-load the libraries/


Load the libraries we are going to use
```{r}
library(DESeq2)
library(readr)
library(RColorBrewer)
library(gplots)
```

Don't forget to set the working directory correctly for your data!
```{r setwd}
#setwd("../data/HTseq_counts")
#setwd("../HTseq_counts")
```
Generate a "data frame" to store the information about each of the samples in your RNA-seq experiment. We could also have done this by creating a CSV or tab-delimited text file with
all of this information and loading it into R with read.csv() or read.table().That would probably be easier, for example, if we had a complicated design with many samples and multiple variables.

```{r sample_names}
samples <- c("ORE_wt_rep1","ORE_wt_rep2","ORE_sdE3_rep1","ORE_sdE3_rep2","SAM_wt_rep1","SAM_wt_rep2","SAM_sdE3_rep1","SAM_sdE3_rep2","HYB_wt_rep1","HYB_wt_rep2","HYB_sdE3_rep1","HYB_sdE3_rep2")

# create names to match file names for the count data
files <- paste(samples, "_htseq_counts.txt", sep="")

backgrounds <- c(rep("ORE", 4), rep("SAM", 4), rep("HYB", 4))

genotypes <- c(rep(c("wt", "wt", "sdE3", "sdE3"), 3))

rna.design <- data.frame(sample = samples, 
                         file = files, 
                         background = backgrounds, 
                         genotype = genotypes)
```

DESeq2 needs a model to load your data; let's create a simple formula to start with, to look for genes that are differentially expressed between two genotypes [wild-type and scalloped-E3 mutants], without accounting for genetic background

```{r}
load.model <- formula(~ genotype)

#Now load the data into R
all.data <- DESeqDataSetFromHTSeqCount(sampleTable=rna.design, directory="./", design=load.model)
```

This next line will handle several steps for us--it will first estimate size factors, to account for differences in library size (total numbers of reads) across samples.Then it will generate dispersion estimates.

```{r}
all.data <- DESeq(all.data)
```

Always a good idea to plot dispersion estimates to make sure they look ok.

```{r}
plotDispEsts(all.data)
```


```{r}
for_pca <- rlog(all.data, blind=TRUE)
dim(for_pca)
```
`rlog` is one approach to adjusting for both library size and dispersion among samples. `blind=TRUE`, has it ignore information from the model (in this case lane).

```{r lane effects, echo=TRUE}
#plotPCA(for_pca, intgroup=c("lane")) 
```

By default this only examine the top 500 genes. Let's look at 2000 or more to get a sense of the stability of the pattern.

```{r lane effects, echo=TRUE}
#plotPCA(for_pca, ntop = 2000, intgroup=c("lane")) 
```


### Back to the analysis
While there is no obvious lane effects based on our initial statistical check, and visual inspection of PC1 VS. PC2. However, there clearly is a pattern here, and it has nothing to do with lane. 

We can quickly take a look to see if this pattern shows anything interesting for our biological interest. However, this is putting the cart before the horse, so be warned.

```{r, echo=TRUE}
plotPCA(for_pca, 
        intgroup=c("genotype", "background"))

plotPCA(for_pca, ntop = 100,
        intgroup=c("genotype", "background"))

plotPCA(for_pca, ntop = 1000,
        intgroup=c("genotype", "background"))
```

This is both heartening and enlightening! Let's think about this for a few minutes and discuss it.

Now let's look at differentially expressed genes

```{r}
genotype.results.df <- as.data.frame(results(all.data, alpha=0.05))
genotype.results <- results(all.data, alpha=0.05)
summary(genotype.results)
```

```{r}
head(genotype.results)
```

Let's sort the genes based on false discovery rate of 0.05

```{r}

genotype.results <- genotype.results[!is.na(genotype.results$padj),]

sig.genotype.results <- genotype.results[order(genotype.results$padj, decreasing=F),]

as.data.frame(sig.genotype.results)[1:80,]
```

Note: FBgn0003345 is *sd*, which is the gene that is mutated in our non-control treatment group in this sample dataset, so it makes sense that it is at the top of our list of differentially expressed genes

We can also make a volcano plot

```{r}
plot(x = genotype.results$log2FoldChange,
     y = -log10(genotype.results$padj), 
     pch = 20, 
     col = ifelse(genotype.results$padj <= 0.05, "red", "black"),
     ylab = "-log10(p)", xlab = "log2 fold change")
```

We really should also account for background (and lane) prior to estimating the genotypic effects. 

```{r more_complex_model}
load.model.2 <- formula(~ background + genotype)

all.data.2 <- DESeqDataSetFromHTSeqCount(sampleTable = rna.design,
                                         directory = "./",
                                         design = load.model.2)

all.data.2 <- DESeq(all.data.2)
```
Let's compare these new dispersion estimates to the old ones

```{r}
#Divide the plot window into two side-by-side panels
par(mfrow=c(1, 2))

#Plot the original dispersion estimates on the left first, then the new dispersion estimates on the right
plotDispEsts(all.data)
plotDispEsts(all.data.2)
```

Although the overall fit looks similar, it's pretty clear that for some genes, the dispersion estimates have changed from using this more complex model

```{r}
genotype.results.2.df <- as.data.frame(results(all.data.2, alpha=0.05))

genotype.results.2 <- results(all.data.2, alpha=0.05)

summary(genotype.results.2)

plotMA(genotype.results.2, ylim =c(-2, 2))

```


```{r}
genotype.results.2 <- genotype.results.2[!is.na(genotype.results.2$padj),]

sig.genotype.2.results <- genotype.results.2[order(genotype.results.2$padj, decreasing=F),]

as.data.frame(sig.genotype.2.results)[1:120,]
```
## Model with interaction
We could also look at models with interaction effects.  (e.g., in this case, which genes are affected by the sdE3 mutation in a background-dependent way?)

```{r interaction_model}
load.model.3 <- formula(~ background + genotype + background:genotype)

all.data.3 <- DESeqDataSetFromHTSeqCount(sampleTable=rna.design, directory="./", design=load.model.3)
all.data.3 <- DESeq(all.data.3)
results.GxB <- results(all.data.3)
summary(results.GxB)
results.GxB <- results.GxB[!is.na(results.GxB$padj), ]
results.GxB <- results.GxB[order(results.GxB$padj),]
as.data.frame(results.GxB)[1:35,]
```

When we have complex models like this and we're interested in comparing two specific treatment groups, we need to use "contrasts". This can get rather complicated. We'll do one or two examples here, but see the DESeq2 tutorial for more details.

### Compare ORE-wt to ORE-sdE3

IGNORE FOR THE MOMENT
```{r}

resultsNames(all.data.3)

contrast.results.3 <- as.data.frame(results(all.data.3, contrast=list(c("backgroundORE.genotypewt", "backgroundORE.genotypewt") ,
						c("genotypesdE3","backgroundORE.genotypesdE3") )))
						
contrast.results.3 <- contrast.results.3[(contrast.results.3$padj <= 0.05) & !is.na(contrast.results.3$padj),]
contrast.results.3 <- contrast.results.3[order(contrast.results.3$padj, decreasing=F),]
```